### DISCUSSION TITLE
Typer command in a python wrapper script

### DISCUSSION THREAD

**Discussion by cousinbr** [fastapi/typer#754] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import sys
from my_typer_app.main import app

sys.stdout = open("batch_file_log", mode="at", encoding="ascii", buffering=1)
sys.stderr = open("batch_file_log", mode="at", encoding="ascii", buffering=1)

app()
print("This line is never executed")
```

### Description

Hello,

I have build a CLI with Typer that do some tasks. I have to schedule task execution (with cron for exemple) but for scheduled task, i want to redirect all output (stdin & stderr) to logfile. SO, i have created a python script that import my Typer app "app" and launch it after stdin/stderr redirection 

Now, i must retrieve "app" return code to send an alert if this is not 0. How could i do that ?

Thanks

### Operating System

Linux

### Typer Version

0.9.0

### Python Version

Python 3.8.16

**Comment by brianmedigate** [comment, marked as answer]
Your script exits before getting to the last print because the `app()` invocation raises a `SystemExit` exception, since it is expecting to be the last thing run in the process.

You could try to catch the exception and access the exit code like this:
```python
try:
    app()
except SystemExit as e:
   print(f"Exited with exit code {e.code}")
```

But a better approach might be to use the `CliRunner` utility from `typer.testing`, which is designed to do basically what you are trying to do: run a typer app from the context of a longer-running piece of code. It can also capture the output for you, which might be cleaner than patching `stderr` and `stdout`.

You can read more about `CliRunner` [here](https://typer.tiangolo.com/tutorial/testing/#import-and-create-a-clirunner) and [here](https://click.palletsprojects.com/en/8.1.x/testing/)

**Reply by YuriiMotov** [reply]
I wouldn't use `CliRunner` for this as it's designed to be used for unit-tests.

The approach with try .. except is working!

```py
import typer

app = typer.Typer()


@app.command()
def create(username: str):
    print(f"Creating user: {username}")


@app.command()
def delete(username: str):
    print(f"Deleting user: {username}")


if __name__ == "__main__":
    try:
        app()
    except SystemExit as e:
        print(f"Exited with exit code {e.code}")
```

Try
```
$ python main.py delete us
Deleting user: us
Exited with exit code 0
```
